### PR TITLE
Change PlotAsymmetryByLogValue To Accept Range Of Runs As Input

### LINKED ISSUE #29160: Change PlotAsymmetryByLogValue to Accept a Range of Runs
Currently only accepts first and last run. Need to accept a range of runs

### PR DESCRIPTION
**Pull request by Matt-Cumber**
**Description of work.**
Added a new array property to PlotAsymmetryByLogValue called WorkspaceNames. Changed the algorithm so that FirstRun and LastRun are no longer mandatory. For successful execution either WorksapceNames must be supplied, or FirstRun and LastRun must be supplied. Added unit tests to test this. 

**To test:**
Run below script and check results are as expected. Results using FirstRun and LastRun as input should be compared to using WorkspaceNames as input, should give the same output. Notice ws_list has only two runs so the output should only be from the first and last run, excluding the middle run. Whereas ws_list_full should be equivalent to using FirstRun and LastRun.
```
first_run = 'MUSR00015189.nxs'
middle_run = 'MUSR00015190.nxs'
last_run = 'MUSR00015191.nxs'

print('####################')
print('Example - Calculating asymmetry for a series of MUSR runs:')
print('####################')

print('First and Last')
ws = PlotAsymmetryByLogValue(FirstRun = first_run,
                             LastRun = last_run,
                             LogValue = 'sample_magn_field',
                             TimeMin = 0.55,
                             TimeMax = 12.0);

print("Y values (asymmetry): {}".format(ws.readY(0)))
print("X values (sample magn. field): {}".format(ws.readX(0)))

ws_list = [first_run, last_run]

print('Short List')
ws = PlotAsymmetryByLogValue(WorkspaceNames = ws_list,
                             LogValue = 'sample_magn_field',
                             TimeMin = 0.55,
                             TimeMax = 12.0);

print("Y values (asymmetry): {}".format(ws.readY(0)))
print("X values (sample magn. field): {}".format(ws.readX(0)))

ws_list_full = [first_run, middle_run, last_run]

print('Full List')
ws = PlotAsymmetryByLogValue(WorkspaceNames = ws_list_full,
                             LogValue = 'sample_magn_field',
                             TimeMin = 0.55,
                             TimeMax = 12.0);

print("Y values (asymmetry): {}".format(ws.readY(0)))
print("X values (sample magn. field): {}".format(ws.readX(0)))

print('####################')
print('Using both Red and Green periods:')
print('####################')

print('First and Last')
ws = PlotAsymmetryByLogValue(FirstRun = 'MUSR00015189.nxs',
                             LastRun = 'MUSR00015191.nxs',
                             LogValue = 'sample_magn_field',
                             TimeMin = 0.55,
                             TimeMax = 12.0,
                             Red = 1,
                             Green = 2);

print("Y values (difference): {}".format(ws.readY(0)))
print("Y values (red): {}".format(ws.readY(1)))
print("Y values (green): {}".format(ws.readY(2)))
print("Y values (sum): {}".format(ws.readY(3)))
print("X values (sample magn. field): {}".format(ws.readX(0)))

print('Short List')
ws = PlotAsymmetryByLogValue(WorkspaceNames = ws_list,
                             LogValue = 'sample_magn_field',
                             TimeMin = 0.55,
                             TimeMax = 12.0,
                             Red = 1,
                             Green = 2);

print("Y values (difference): {}".format(ws.readY(0)))
print("Y values (red): {}".format(ws.readY(1)))
print("Y values (green): {}".format(ws.readY(2)))
print("Y values (sum): {}".format(ws.readY(3)))
print("X values (sample magn. field): {}".format(ws.readX(0)))

print('Full List')
ws = PlotAsymmetryByLogValue(WorkspaceNames = ws_list_full,
                             LogValue = 'sample_magn_field',
                             TimeMin = 0.55,
                             TimeMax = 12.0,
                             Red = 1,
                             Green = 2);

print("Y values (difference): {}".format(ws.readY(0)))
print("Y values (red): {}".format(ws.readY(1)))
print("Y values (green): {}".format(ws.readY(2)))
print("Y values (sum): {}".format(ws.readY(3)))
print("X values (sample magn. field): {}".format(ws.readX(0)))
```

**To test the dialog window:** 

1. Set your default instrument to **MUSR**
2. Open the dialog window for **PlotAsymmetryByLogValue** from the algorithms widget on mantid workbench
3. In the runs box enter `15189-15191`. This will load files **MUSR15189, MUSR15190, MUSR15191**
4. Remember to enter a name for the **Output workspace**
5. Show data on the output workspace and hopefully you have something that looks as follows:

image

6. Feel free to test different ranges to check no unexpected errors occur, use commas to include individual runs e.g. `15189-15191,15195`

Fixes #29160 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by AnthonyLim23 — DISMISSED** [review]
Could we update the documentation to say preference is given to the list and add an example of using a list.

File: docs/source/release/v5.1.0/framework.rst [hunk 1]
```diff
@@ -48,6 +48,9 @@ Algorithms
 - Add parameters to :ref:`LoadSampleShape <algm-LoadSampleShape>` to allow the mesh in the input file to be rotated and\or translated
 - Algorithms now lazily load their documentation and function signatures, improving import times from the `simpleapi`.
 - Deprecated the RecordPythonScript algorithm
+- Add parameter 'WorkspaceNames' to :ref:`PlotAsymmetryByLogValue <algm-PlotAsymmetryByLogValue>` which accepts a range of runs to be processed.
```
Comment (StephenSmith25) [inline docs/source/release/v5.1.0/framework.rst:51]: I'm wondering whether framework is the best place to document these changes. I think putting them in the Muon docs would be more appropriate, @AnthonyLim23  thoughts?

**Review by StephenSmith25 — CHANGES_REQUESTED** [review]
Generally looks very good and works well 👍, just a couple of comments - One thing to note is I don't think this new option appears in the algorithm dialog window (when you search for the algorithm in workbench)

**Comment by Matt-Cumber** [comment]
> Generally looks very good and works well 👍, just a couple of comments - One thing to note is I don't think this new option appears in the algorithm dialog window (when you search for the algorithm in workbench)

@StephenSmith25, So this was an issue which came from #24839 to change the ALC interface, but it was noticed that the interface relies on this algorithm so I had to make the changes in this PR first. @AnthonyLim23, is it necessary to change the dialog window as well, and is that something I also have to do? and would it be part of this issue or a new one?

File: Framework/Muon/inc/MantidMuon/PlotAsymmetryByLogValue.h [hunk 2]
```diff
@@ -116,6 +118,12 @@ class MANTID_MUON_DLL PlotAsymmetryByLogValue : public API::Algorithm {
   std::vector<int> m_forward_list;
   /// Store backward spectra
   std::vector<int> m_backward_list;
+  /// Store workspaces
+  std::vector<std::string> m_fileNames;
+  /// A typedef for the map of workspace names to run numbers
+  using RunNumberMap = std::map<std::string, int>;
```
[the remark this answers is not in the document]
Comment (Matt-Cumber) [reply Framework/Muon/inc/MantidMuon/PlotAsymmetryByLogValue.h:124]: I guess that makes sense, my reasoning for doing this was to have a more meaningful type name. I can remove the typedef

File: Framework/Muon/src/PlotAsymmetryByLogValue.cpp [hunk 3]
```diff
@@ -234,17 +260,45 @@ void PlotAsymmetryByLogValue::checkProperties(size_t &is, size_t &ie) {
...
+    is = std::stoul(firstFN); // starting run number
+    ie = std::stoul(lastFN);  // last run number
+    if (ie < is) {
+      throw std::runtime_error(
+          "First run number is greater than last run number");
+    }
+
+    for (size_t i = is; i <= ie; i++) {
+      // Get complete run name
+      std::ostringstream fn, fnn;
```
[the remark this answers is not in the document]
Comment (Matt-Cumber) [reply Framework/Muon/src/PlotAsymmetryByLogValue.cpp:279]: Yes you are right I just copied what was there before, I've changed the names

File: docs/source/release/v5.1.0/framework.rst (hunk 1 above)
Comment (AnthonyLim23) [reply docs/source/release/v5.1.0/framework.rst:51]: These are muon specific, the other instruments don't use asymmetry

**Comment by AnthonyLim23** [comment]
We have 3 choices:
1. Expose both runs, first & last to the algorithm dialog
2. Expose the runs
3. Expose the first & last (what we currently have)

I would be inclined to say 2 is better as we can replicate the behaviour of 3, but 3 cannot replicate the behaviour of 2.

File: Framework/Muon/src/PlotAsymmetryByLogValue.cpp [hunk 4]
```diff
@@ -234,17 +260,45 @@ void PlotAsymmetryByLogValue::checkProperties(size_t &is, size_t &ie) {
...
   std::string firstFN = getProperty("FirstRun");
   std::string lastFN = getProperty("LastRun");
+  m_fileNames = getProperty("WorkspaceNames");
+
+  // If file names empty, first and last provided so need to populate vector
+  if (m_fileNames.empty()) {
+    // Parse run names and get the number of runs
+    parseRunNames(firstFN, lastFN, m_filenameBase, m_filenameExt,
+                  m_filenameZeros);
+    is = std::stoul(firstFN); // starting run number
```
Comment (StephenSmith25) [inline Framework/Muon/src/PlotAsymmetryByLogValue.cpp:270]: Both `is` and `ie` would probably benefit from better names

**Review by StephenSmith25 — COMMENTED** [review]
Just one final suggestion. Also, I'm having some trouble working with the newly exposed runs option in the algorithm dialog. Maybe I'm missing something but it doesn't seem clear how I select a series of runs using it.

**Comment by Matt-Cumber** [comment]
> Just one final suggestion. Also, I'm having some trouble working with the newly exposed runs option in the algorithm dialog. Maybe I'm missing something but it doesn't seem clear how I select a series of runs using it.

Assume MUSR is the default instrument. A series of runs can be entered like 

> Just one final suggestion. Also, I'm having some trouble working with the newly exposed runs option in the algorithm dialog. Maybe I'm missing something but it doesn't seem clear how I select a series of runs using it.

@StephenSmith25 I have updated the testing instructions to include how to test the dialog window. Let me know if you have any more trouble, thanks for taking the time to review this.

**Comment by StephenSmith25** [comment]
Great thanks for updating the testing instructions. Code changes look good overall. I still find it a bit awkward to use - for instance typing in the box and clicking run doesn't load the correct list - to update it you have to click somewhere else in the dialog and wait. This may just be inherent in the design so likely outside the scope of the PR. Also, it appears to me that the browse button seems to lose its purpose now, you can use it to load a single file, but when you load another one it will clear the runs box.

**Comment by AnthonyLim23** [comment]
> Great thanks for updating the testing instructions. Code changes look good overall. I still find it a bit awkward to use - for instance typing in the box and clicking run doesn't load the correct list - to update it you have to click somewhere else in the dialog and wait. This may just be inherent in the design so likely outside the scope of the PR. Also, it appears to me that the browse button seems to lose its purpose now, you can use it to load a single file, but when you load another one it will clear the runs box.

You should be able to have multiple files from browse by holding down control (I think, it might be shift)

**Comment by Matt-Cumber** [comment]
> Great thanks for updating the testing instructions. Code changes look good overall. I still find it a bit awkward to use - for instance typing in the box and clicking run doesn't load the correct list - to update it you have to click somewhere else in the dialog and wait. This may just be inherent in the design so likely outside the scope of the PR. Also, it appears to me that the browse button seems to lose its purpose now, you can use it to load a single file, but when you load another one it will clear the runs box.

Yes as @AnthonyLim23 mentioned holding **Control** in the browse window to select multiple files then clicking open will load in multiple files to the runs box. As for the runs box updating. It looks like if you make changes and press **Enter** the data is not updated, but if you make changes and manually click the **Run** button the data updates fine. Not entirely sure why that is.

**Review by StephenSmith25 — APPROVED** [review]
Looks good to me 👍

**Pull request merged**